src/i2p.cpp

```diff
@@ -0,0 +1,432 @@
+{
+    Reset();
+    m_socket = other.m_socket;
+    other.m_socket = INVALID_SOCKET;
+}
+
+Sock::~Sock() { Reset(); }
+
+Sock& Sock::operator=(Sock&& other)
+{
+    Reset();
+    m_socket = other.m_socket;
+    other.m_socket = INVALID_SOCKET;
+    return *this;
+}
+
+SOCKET Sock::Get() const { return m_socket; }
+
+SOCKET Sock::Release()
+{
+    const SOCKET s = m_socket;
+    m_socket = INVALID_SOCKET;
+    return s;
+}
+
+void Sock::Reset() { CloseSocket(m_socket); }
+
+Session::Session(const fs::path& private_key_file,
+                 const CService& control_host,
+                 CThreadInterrupt* interrupt)
+    : m_private_key_file(private_key_file), m_control_host(control_host), m_interrupt(interrupt),
+      m_control_sock(INVALID_SOCKET)
+{
+}
+
+Session::~Session()
+{
+    Disconnect();
+}
+
+bool Session::Accept(Connection& accepted)
+{
+    static constexpr auto err_wait_begin = 1s;
+    static constexpr auto err_wait_cap = 1min;
+    auto err_wait = err_wait_begin;
+
+    while (!*m_interrupt) {
+        try {
+            Sock sock;
+
+            {
+                LOCK(m_mutex);
+                CreateIfNotCreatedAlready();
+                accepted.me = m_my_addr;
+                sock = StreamAccept();
+            }
+
+            while (!*m_interrupt) {
+                Sockets sockets;
+                sockets.recv.insert({sock.Get()});
+                sockets.err.insert({sock.Get()});
+
+                SocketEvents(sockets, MAX_WAIT_FOR_IO, *m_interrupt);
+
+                if (sockets.recv.count(sock.Get()) == 0 && sockets.err.count(sock.Get()) == 0) {
+                    // No incoming connections within MAX_WAIT_FOR_IO.
+                    continue;
+                }
+
+                const std::string& peer_dest =
+                    RecvUntilTerminator(sock.Get(), '\n', MAX_WAIT_FOR_IO, *m_interrupt);
+
+                accepted.peer = CService(DestB64ToAddr(peer_dest), Params().GetDefaultPort());
+                accepted.socket = sock.Release();
+
+                err_wait = err_wait_begin;
+
+                return true;
+            }
+        } catch (const std::runtime_error& e) {
+            Log("Error accepting: %s", e.what());
+            m_interrupt->sleep_for(err_wait);
+            if (err_wait < err_wait_cap) {
+                err_wait *= 2;
+            }
+        }
+    }
+    return false;
+}
+
+bool Session::Connect(const CService& to, Connection& conn, bool& proxy_error)
+{
+    proxy_error = true;
+
+    std::string session_id;
+    Sock sock;
+    conn.peer = to;
+
+    try {
+        {
+            LOCK(m_mutex);
+            CreateIfNotCreatedAlready();
+            session_id = m_session_id;
+            conn.me = m_my_addr;
+            sock = Hello();
+        }
+
+        const Reply& lookup_reply =
+            SendRequestAndGetReply(sock, strprintf("NAMING LOOKUP NAME=%s", to.ToStringIP()));
+
+        const std::string& dest = lookup_reply.Get("VALUE");
```

Comment:
What if no `VALUE` is received?

Reply:
Then we will get an exception which is properly handled. Do you see any problems?

Reply:
No, I don't see any problem, it is just that I didn't see the exception handling.

Reply:
Ok :)